gdk/commands/component/recipe_generator/BuildRecipeGenerator.py

```diff
@@ -0,0 +1,137 @@
+import logging
+import shutil
+from pathlib import Path
+from gdk.common.CaseInsensitive import CaseInsensitiveRecipeFile, CaseInsensitiveDict
+
+import gdk.commands.component.project_utils as project_utils
+import gdk.common.consts as consts
+import gdk.common.utils as utils
+
+
+class BuildRecipeGenerator:
+    def __init__(self, project_config) -> None:
+        self.project_config = project_config
+
+    def generate(self, build_folders):
+        component_recipe = CaseInsensitiveRecipeFile().read(self.project_config["component_recipe_file"])
+        self.update_component_recipe_file(component_recipe, build_folders)
+        self.create_build_recipe_file(component_recipe)
+
+    def update_component_recipe_file(self, parsed_component_recipe: CaseInsensitiveDict, build_folders):
+        logging.debug(
+            "Updating component recipe with the 'component' configuration provided in '{}'.".format(
+                consts.cli_project_config_file
+            )
+        )
+        parsed_component_recipe.update_value("ComponentName", self.project_config["component_name"])
+        parsed_component_recipe.update_value("ComponentVersion", self.project_config["component_version"])
+        parsed_component_recipe.update_value("ComponentPublisher", self.project_config["component_author"])
+        self.update_artifact_uris(parsed_component_recipe, build_folders)
+
+    def update_artifact_uris(self, parsed_component_recipe, build_folders: list) -> None:
+        """
+        The artifact URIs in the recipe are used to identify the artifacts in local build folders of the component or on s3.
+
+        If the artifact is not found in the local build folders specific to the build system of the component, it is
+        searched on S3 with exact URI in the recipe.
+
+        Build command fails when the artifacts are neither not found in local both folders not on s3.
+        """
+        logging.info("Copying over the build artifacts to the greengrass component artifacts build folder.")
+        logging.info("Updating artifact URIs in the recipe.")
+        if "Manifests" not in parsed_component_recipe:
+            logging.debug("No 'Manifests' key in the recipe.")
+            return
+        manifests = parsed_component_recipe["Manifests"]
+        s3_client = None
+        for manifest in manifests:
```

Comment:
All this nested conditionals here are a code smell. The `parsed_component_recipe` is a misleading, when we parse data we should be converting the data from one format to a format we know is already valid, otherwise passing a CaseInsensitiveDict dict around forces the received to always have to do validations 

Here is an article that explains the mantra of parse, not validate https://towardsdatascience.com/parse-dont-validate-f559372cca45

Reply:
If you seen the code below the other "Generator" is having to do similar validation all over again because we are not properly parsing and we are relying to much on the dictionary.